src/lib.rs

```diff
@@ -1311,6 +1315,112 @@ where
     Some(Ratio::new(n1, d1))
 }
 
+#[cfg(all(feature = "bigint", feature = "std"))]
+impl<T: Clone + Integer + Signed + ToPrimitive + ToBigInt> Ratio<T> {
+    /// Converts the ratio to an `f64`.
+    ///
+    /// The value is guaranteed to be the closest `f64` approximating the ratio. This always
+    /// returns a value, unless the conversion to BigInt for the numerator or denominator fails.
+    ///
+    /// ```
+    /// use num_rational::Rational64;
+    /// assert_eq!(0.5f64, Rational64::new(1, 2).to_f64().unwrap());
+    /// ```
+    pub fn to_f64(&self) -> Option<f64> {
+        assert_eq!(
+            std::f64::RADIX,
+            2,
+            "only floating point implementations with radix 2 are supported"
+        );
+
+        // Upper bound to the range of exactly-representable ints in an f64.
+        const MAX_EXACT_INT: u64 = 1u64 << std::f64::MANTISSA_DIGITS;
+
+        let numer: T = self.numer().abs();
+        let denom: T = self.denom().abs();
+        let sign: T = self.numer().signum() * self.denom().signum();
+        let flo_sign = if sign.is_negative() { -1.0 } else { 1.0 };
+
+        if numer.is_zero() {
+            return Some(0.0 * flo_sign);
+        }
+
+        // Fast track: both sides can losslessly be converted to f64s. In this case, letting the
+        // FPU do the job is faster and easier. In any other case, converting to f64s may lead
+        // to an inexact result: https://stackoverflow.com/questions/56641441/.
+        if let (Some(n), Some(d)) = (numer.to_u64(), denom.to_u64()) {
+            if n < MAX_EXACT_INT && d < MAX_EXACT_INT {
+                return Some(numer.to_f64().unwrap() / denom.to_f64().unwrap() * flo_sign);
```

Comment:
this should be `return Some(flo_sign * n / d)`

Reply:
Do you mean `Some(flo_sign * n as f64 / d as f64)`? I think cast on line 1351 should be to `u64`, as it currently is, so I need to cast to float here. (Actually, as the code is currently written, the first cast could be directly to `f64`, but the inequalities on line 1352 should be non-strict as `MAX_EXACT_INT` is in fact exactly representable). This is because casting `MAX_EXACT_INT + 1` as an `f64` will return `MAX_EXACT_INT` as well.

Reply:
nevermind, I misread your code. It could be written differently to avoid `unwrap()`, which is all I was trying to do.

Reply:
So are you OK with the current `as` cast, or would you prefer something else? I'm pretty agnostic about this.

Reply:
I wanted to avoid `unwrap()` because library code ideally shouldn't panic. Using `as` is a step in the wrong direction, as it will silently fail (I think?), giving a garbage number if the arguments are out of range ([like here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=2dec23c6d13520817b8d498e1b40e8d0)). I was thinking something more like
```rust
if let (Some(n), Some(d)) = (numer.to_u64(), denom.to_u64()) {
    if n < MAX_EXACT_INT && d < MAX_EXACT_INT {
        if let (Some(n), Some(d)) = (numer.to_f64(), denom.to_f64()) {
            return Some(flo_sign * n / d);
        }
    }
}
```
This is statically guaranteed to not panic, whereas what you wrote is "dynamically guaranteed" to not panic (since f64's `MAX_EXACT_INT` couldn't possibly be higher than the max value of an `i64` or `u64`).

Reply:
>  Using `as` is a step in the wrong direction, as it will silently fail

That's true in general, but `u64`→`f64` is [guaranteed to succeed](https://doc.rust-lang.org/reference/expressions/operator-expr.html#type-cast-expressions). I agree that that's not really a static proof. Normally you can use the `From` trait for a static proof, but it looks like it's not implemented for this cast as not all `u64` are exactly representable as `f64`s.

At a higher level, I agree that, as much as possible, you want to statically prove that the algorithm works by encoding that in the type system. However, if that's not possible, I think it's fine for library code to panic *when it detects a bug*, so to speak. If the algorithm is in a state that should be unreachable, it's better to panic early than to carry on and return a wrong result or at least confuse the person trying to debug it.

I think the [Rust book agrees](https://doc.rust-lang.org/book/ch09-03-to-panic-or-not-to-panic.html), although it does focus more on incorrect user input than incorrect internal state:

> It would also be appropriate to call `unwrap` when you have some other logic that ensures the `Result` will have an `Ok` value, but the logic isn’t something the compiler understands.

In this particular case, the inner `if let` in the code you propose will either always match, or the algorithm is broken. By using `if let`, we're introducing the possibility of hiding invalid state. Since it doesn't seem possibly to statically prove the conversion succeeds, I think this is a situation where `unwrap()` is more appropriate.